Improved Gauging Interface, TNO construction, Bug Fixes, Apply Function for a Vidal ITN

JoeyT1994 opened the pull request:
This PR adds a number of features in ITensorNetworks.

1) The gauging interface has been expanded significantly (via `gauging.jl`). Functions are now available for belief propagation gauging a tensornetwork, starting in either symmetric or Vidal form. Functions are available for switching between these two forms, and for measuring the degree of `canonicalness' (i.e how well the itn satisfied certain canonical conditions) of a given itensornetwork (either in Vidal or Symmetric form).
2) An additional `apply(o::ITensor, psi::ITensornetwork, bond_tensors::Datagraph)` function has been added to `apply.jl` which performs a simple update on a tensornetwork in the Vidal gauge. Ideally this will eventually share more code with `apply(o::ITensor, psi::ITensornetwork)` but for now I have left them more separate.
3) Functionality has now been added (see `tensornetworkoperators.jl`) for grouping up the `ITensors` in a `vector{ITensor}`object into groups of commuting `ITensors` (i.e. ones that act on different physical degrees of freedom). This is the workhorse of a `get_tnos` function which builds a vector of `tensornetworkoperator`s from a vector of ITensors.
4) A few tweaks have been made in `beliefpropagation.jl` to improve efficiency and avoid excessive calls to forming the union of two itensornetworks (which can be quite expensive).
5) Extensive testing has been added for the above, including additions to tests of the `apply` function

mtfishman commented on src/gauging.jl:
```diff
@@ -74,21 +70,21 @@ function vidal_gauge(
   svd_kwargs...,
 )
-  bond_tensors = DataGraph{vertextype(ψ),ITensor,ITensor}(underlying_graph(ψ))
+  bond_tensors = intialise_bond_tensors(ψ)
```
`intialise` to `initialize`.

mtfishman replied:
Apparently "initialise" is the British spelling, I didn't know that! https://english.stackexchange.com/questions/459553/initialised-or-initialized-which-one-is-correct-spelling.

JoeyT1994 replied:
Yeah it is! Similarly with colour vs color and neighbour vs neighbor. Makes sense to use the American spelling though - might confuse people otherwise. Just think I am on autopilot when I name things

mtfishman commented on src/ITensorsExt/itensorutils.jl:
```diff
@@ -19,3 +19,32 @@ inv_diag(it::ITensor) = map_diag(inv, it)
 pinvsqrt_diag(it::ITensor) = map_diag(pinv ∘ sqrt, it)
+
+"""Given a vector of ITensors, separate them into groups of commuting itensors (i.e. itensors in the same group do not share any common indices)"""
+function group_ITensors(its::Vector{ITensor})
```
Change to all lower case, i.e. `group_itensors`. The convention in Julia is that function names are all lower case. Only types, constructors, and module names have capital letters.

Let's think of a more specific name for this, for example something that specifies _how_ the ITensors are being grouped, since they could be grouped in many different ways.

mtfishman commented on src/apply.jl:
```diff
@@ -1,14 +1,13 @@
   nfullupdatesweeps=10,
   print_fidelity_loss=true,
   envisposdef=false,
+  svd_kwargs...,
```
I would call this `apply_kwargs` since it is getting passed to the `apply` function.

mtfishman commented on src/apply.jl:
```diff
@@ -87,52 +86,124 @@ end
+)
+  ψ = copy(ψ)
+  bond_tensors = copy(bond_tensors)
+  v⃗ = typeof(o) == ITensor ? neighbor_vertices(ψ, o) : [src(o), dst(o)]
```
I try to avoid code like this that uses `if` statements based on checking types. I find it makes the code cleaner to instead define a function and then dispatch on that function.

So for example:
```julia
_gate_vertices(o::ITensor, ψ) = neighbor_vertices(ψ, o)
_gate_vertices(o::AbstractEdge, ψ) = [src(o), dst(o)]
```
and then use:
```julia
  v⃗ = _gate_vertices(o, ψ)
```

mtfishman commented on src/apply.jl:
```diff
@@ -87,52 +86,124 @@ end
+      end
+    end
+
+    if typeof(o) == ITensor
+      G1, G2 = factorize(o, Index[commonind(ψv1, o), commonind(ψv1, o)']; cutoff=1e-16)
+      ψv1 = noprime(ψv1 * G1)
+      ψv2 = noprime(ψv2 * G2)
+    end
```
Same comment as above, I would split this off into a separate function like:
```julia
function _contract_factorized_gate(o::ITensor, ψv1, ψv2)
  G1, G2 = factorize(o, Index[commonind(ψv1, o), commonind(ψv1, o)']; cutoff=1e-16)
  ψv1 = noprime(ψv1 * G1)
  ψv2 = noprime(ψv2 * G2)
  return ψv1, ψv2
end

function _contract_factorized_gate(o::AbstractEdge, ψv1, ψv2)
  return ψv1, ψv2
end
```
I think splitting off small functions like this helps "self-document" the code and helps to identify patterns for how the code could be generalized (for example, for other gate types, like gates that are MPOs or TTNOs).

JoeyT1994 replied:
Yeah I can see that that is a neater way to write the code! Thanks

JoeyT1994 commented:
Okay I incorporated the changes you suggested above! Do you know what is going on with the checks? Suddenly, for some reason certain packages aren't compiling? Even though I don't think I have changed the manifest/ dependencies.

mtfishman commented:
Looks like https://github.com/SciML/LinearSolve.jl/issues/330.

JoeyT1994 commented:
I see! That makes sense.

mtfishman commented:
We don't have any control over that issue, so as long as things pass in Julia 1.9 I think it's fine to merge this.

mtfishman commented:
Looks good, thanks! This will be very useful to have, I'm sure we'll find many use cases for it.

emstoudenmire commented:
Great to see this, thanks Joey and Matt!

On Fri, Jun 16, 2023 at 2:37 PM Matt Fishman ***@***.***>
wrote:

> Merged #88 <https://github.com/mtfishman/ITensorNetworks.jl/pull/88> into
> main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mtfishman/ITensorNetworks.jl/pull/88#event-9554645617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHJ3RLDVRKUFXDWVMUW5VTXLSRVNANCNFSM6AAAAAAYCM2D2M>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>
-- 
   -=Miles Stoudenmire=-
   ***@***.***
   ***@***.***
   http://itensor.org/miles/
